lib/processExportsField.js

```diff
@@ -0,0 +1,270 @@
+		const export_ = exportsFieldDefinition[exports[i]];
+
+		if (isConditionalMapping(export_)) {
+			return conditionalMapping(
+				remainingRequest,
+				expectFolder,
+				/** @type {ConditionalMapping} */ (export_),
+				conditionNames
+			);
+		}
+
+		return directMapping(
+			remainingRequest,
+			expectFolder,
+			/** @type {string|string[]} */ (export_)
+		);
+	}
+
+	return [];
+};
+
+/**
+ * @param {string} request request
+ */
+function assertRequest(request) {
+	if (request.charCodeAt(0) !== dotCode) {
+		throw new Error('Request should be relative path and start with "."');
+	}
+	if (request.length === 1) return;
+	if (request.charCodeAt(1) !== slashCode) {
+		throw new Error('Request should be relative path and start with "./"');
+	}
+	if (request.charCodeAt(request.length - 1) === slashCode) {
+		throw new Error("Only requesting file allowed");
+	}
+}
+
+/**
+ * Trying to match request to export field
+ * @param {string} request request
+ * @param {string} exportFieldKey export field key
+ * @returns {number} if found match return remaining request start index, else -1
+ */
+function tryToMatch(request, exportFieldKey) {
+	if (exportFieldKey.charCodeAt(0) !== dotCode) {
+		throw new Error(
+			'Export field key should be relative path and start with "."'
+		);
+	}
+
+	if (request.length === 1) {
+		return exportFieldKey.length === 1 ? 1 : -1;
+	}
+
+	if (exportFieldKey.charCodeAt(1) !== slashCode) {
+		throw new Error(
+			'Export field key should be relative path and start with "./"'
+		);
+	}
+
+	let i = 2;
+
+	while (i < exportFieldKey.length) {
+		if (exportFieldKey.charCodeAt(i) !== request.charCodeAt(i++)) return -1;
+	}
+
+	// file mapping
+	if (request.length === i) {
+		return i;
+	}
+
+	return request.charCodeAt(i - 1) === slashCode ? i : -1;
+}
+
+/**
+ * @param {string} exp export target
+ * @param {boolean} expectFolder is folder expected
+ */
+function assertExport(exp, expectFolder) {
+	if (
+		exp.charCodeAt(0) === slashCode ||
+		(exp.charCodeAt(0) === dotCode && exp.charCodeAt(1) !== slashCode)
+	) {
+		throw new Error('Export should be relative path and start with "./"');
+	}
+
+	const isFolder = exp.charCodeAt(exp.length - 1) === slashCode;
+
+	if (isFolder !== expectFolder) {
+		throw new Error(
+			expectFolder
+				? "Expecting folder to folder mapping. Folder to file found"
+				: "Expecting file to file mapping. File to folder found"
+		);
+	}
+}
+
+/**
+ * @param {ConditionalMapping|Mapping} mapping mapping
+ * @returns {boolean} is conditional mapping
+ */
+function isConditionalMapping(mapping) {
+	return typeof mapping === "object" && !Array.isArray(mapping);
+}
+
+/**
+ * @param {string} remainingRequest remaining request
+ * @param {boolean} expectFolder expect folder
+ * @param {string|string[]} directExport direct export
+ * @returns {string[]} mapping result
+ */
+function directMapping(remainingRequest, expectFolder, directExport) {
+	const isString = typeof directExport === "string";
```

Comment:
Make a big `if (typeof directExport === "string") else` instead of storing it in a variable and testing it multiple times. This improves type inference and requires less `?:` here. (Less branches, better performance, but probably doesn't really matter here, is more for readability)